build_tools/python/e2e_test_framework/definitions/common_definitions.py

```diff
@@ -56,15 +56,16 @@ class DeviceArchitecture(Enum):
 
 @dataclass(frozen=True)
 class PlatformInfo(object):
-  """Platform information of a device."""
+  """Platform information of a host."""
   os: str
+  architecture: str
```

Comment:
In https://github.com/iree-org/iree/pull/11413 we just made platform not depend on arch. Let's be consistent (I guess you authored this before the changes from PR review).

Reply:
I don't have a good idea to name the platform+arch. I call it `HostEnvironment` here. Let me know if you have better ideas.

Reply:
Ugh now that I'm looking at VMVX being listed as an architecture, I'm questioning our decision again. IMO "platform" is a generic term indicating where a piece of software is running and what it has access to. It could be a baremetal hardware arch with no OS or similar supporting it, a virtual machine like the JVM or VMVX, an operating system abstracting over the underlying hardware, an operating system and the underlying architecture, etc. @benvanik we were keying off of [target_platform.h](https://github.com/iree-org/iree/blob/main/runtime/src/iree/base/target_platform.h), which lists arch separately from platform, but I think the platform can include that as well. It also lists WASM as an architecture, which doesn't seem quite right. Ugh all of these are loosely used terms. I think the way Bazel defines this is actually pretty reasonable: https://bazel.build/extending/platforms, although maybe I would actually reverse it. I don't know where that leaves us in terms of C macros though...

@ScottTodd do you have thoughts?

Reply:
Wondering if we can see VMVX, WASM as architectures (they also have instruction sets). I guess the virtual machine to run VMVX is also called VMVX right now, so `platform: VMVX, arch: VMVX` seems okay? Same for WASM (and there are multiple VM implementations for WASM, so we use different platforms for them).

The actual host that runs the virtual machine still can affect the performance, but it looks like another layer in the system. And that layer can have its own platform-arch variants (like VMVX virtual machines running on Android-x86_64).

Reply:
So I think the `architecture` here means the `architecture` of runtime executable itself (not the architecture of the module running on it). It also doesn't specify which runtime should be used here.

So for a VMVX benchmark, if we say run it on an `platform:Linux, architecture: x86_64` host, this means:
1. The runtime/tool should be a Linux x86_64 binary
2. The runtime/tool needs to support VMVX ISA

The 2. is not specified here (actually nowhere currently) since our benchmark tool is built to support all ISA.

Reply:
[target_platform.h](https://github.com/iree-org/iree/blob/main/runtime/src/iree/base/target_platform.h) lists WASM as architecture seems right to me. WASM is an ISA, and we can have different WASM VM as platform to run an IREE runtime that is compiled in WASM

Reply:
> So I think the architecture here means the architecture of runtime executable itself (not the architecture of the module running on it).

Ah good point. In that case, I think having VMVX as an architecture is a mistake.

> The runtime/tool needs to support VMVX ISA

Right. I guess the fact that the module is targeting VMVX is something specified in the benchmark definition, not the platform definition.

> WASM as architecture also seems right. WASM is an ISA, and we can have different WASM VMs as platforms to run an IREE runtime that is compiled into WASM

Right, ok, so we're using "architecture" here to mean "instruction set architecture", specifically. I think that's consistent with LLVM's usage also, where wasm64 is an arch (https://llvm.org/doxygen/classllvm_1_1Triple.html). If we want to be consistent with that (and target_platfrom.h) then the arm specializations (v8.2, v9, etc) should be part of the microarchitecture. We should also maybe use the official name AArch64 and the retronym AArch32.

But I think my point is that arch can be a component of the "platform" which describes the entire environment in which a piece of software is executing and therefore what features it can expect to use: https://en.wikipedia.org/wiki/Computing_platform. If the software were running entirely in a VM though and didn't have access to the ISA instructions directly, then it wouldn't be a component of that though (and we shouldn't be compiling it in a way that the compilation knows about it, probably). So I think target_platform.h is "wrong" in the way that it makes platform separate from arch (in quotes because all of these are under-defined concepts full of debate). Unless it's intended that the platform macros are for components of the platform that we don't have a specific category for. Notably, right now, multiple platform macros can be true, e.g. `IREE_PLATFORM_IOS` implies `IREE_PLATFORM_APPLE`. This makes me think that they're really more like constraints/capabilities. Some of them are mutually exclusive, but they're not in general. I guess for ones that have well-defined groupings it would make sense to give them macros with the same name prefix. We do this for architecture, endianness, and memory alignment already (compiler is different, since that's not part of the runtime). We could have `IREE_OS_FAMILY_*` macros for `LINUX`, `WINDOWS`, and `MACOS` and maybe even the OS "super families" like `APPLE`. Then we have `IREE_PLATFORM_EMSCRIPTEN`, which I don't really understand, since I think that's really a compiler. `IREE_PLATFORM_ANDROID_EMULATOR` doesn't appear to ever be set or used and `IREE_PLATFORM_IOS_SIMULATOR` is set but never used, so we could delete those and punt on further categorization.

Maybe I've talked myself into saying that the `IREE_PLATFORM_*` macros are fine if we interpret them to mean `IREE_PLATFORM_COMPONENT_*`. The Bazel "constraint" naming doesn't seem quite right, since a platform offers both constraints and capabilities.

Turning back to our benchmark stuff... I think we just don't want to back ourselves into a corner. Could platform include both structured components (architecture, microarchitecture, os family, os) and unstructured components for things that don't fit neatly into any of the categories?

Reply:
On the benchmark side, I'm fine to switch back to the flexible `<os/platform>-<architecture>` format for `platform`. I don't know if we want to invent a better format, like `<os-family>-<platform>-<architecture>-<microarchitecture>` here. The `<os/platform>-<architecture>` seems to be enough for now.

For the `<architecture>` for ARM, I was following #9021 (https://community.arm.com/arm-community-blogs/b/tools-software-ides-blog/posts/compiler-flags-across-architectures-march-mtune-and-mcpu) and consider `armv8.2-a` as `<architecture>` and something like `cortex-a72` as `<microarchitecture>` (essentially architecture is ISA and microarchitecture is the variants of hardware). `armv8.2-a` is also the accepted name on LLVM's `-march` option. Not sure if we want to invent the `aarch64-v8.2-a` by following the LLVM enums.

Reply:
Hmm weird that the target triple defines it differently. I guess `SubArchType` isn't the same as "microarchitecture" as I had assumed?

So VMVX is an architecture in the device architecture, which makes sense because that *is* about what the module is targeting. But it's not an architecture in the host environment because the IREE runtime is not itself running on VMVX. I think this makes sense. I think it probably also makes sense to be consistent with target_platform.h even if we should maybe change that. Maybe could use some more documentation though :-)

Reply:
SG. Added more explanation.